File: sylee723/3주차/PGS_다리를_지나는_트럭.java
import java.util.*;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        ArrayDeque<Truck> bridge = new ArrayDeque<>();
        int time = 1;
        int total = 0;
        for (int i = 0; i < truck_weights.length; i++) {
            while ((!bridge.isEmpty() && bridge.peek().arriveTime <= time) || (total + truck_weights[i] > weight)) {
                Truck arrive = bridge.poll();
                total -= arrive.weight;
                time = arrive.arriveTime;
            }
            
            total += truck_weights[i];
            bridge.add(new Truck(truck_weights[i], time + bridge_length));
            time++;        
        }
        
        int answer = bridge.peekLast().arriveTime;
        return answer;
    }
    
    static class Truck {
        int weight, arriveTime;
        
        public Truck(int weight, int arriveTime) {
            this.weight = weight;
            this.arriveTime = arriveTime;
        }
    }
}

File: daehwa_park/10주차/PGS_다리를_지나는_트럭.py
from collections import deque

def solution(bridge_length, weight, truck_weights):
    time = 0
    bridge_w = 0
    
    q = deque()
    
    while True:
        if truck_weights and bridge_w + truck_weights[0] <= weight:
            cur = truck_weights.pop(0)
            q.append([cur, 0])
            bridge_w += cur
        for c in q:
            c[1] += 1
        if q[0][1] == bridge_length:
            w, cnt = q.popleft()
            bridge_w -= w
        time += 1
        if not q and not truck_weights:
            break
    
    return time + 1